src/psyclone/psyir/nodes/omp_directives.py

```diff
@@ -227,6 +237,814 @@ class OMPSerialDirective(OMPRegionDirective, metaclass=abc.ABCMeta):
+            task2 = pair[1]
+
+            # Keep track of if the dependencies are satisfiable
+            satisfiable = True
+            # Find all References in each tasks' depend clauses
+            # Should we cache these instead?
+            task1_in = [x for x in task1.children[4].children
+                        if isinstance(x, Reference)]
+            task1_out = [x for x in task1.children[5].children
+                         if isinstance(x, Reference)]
+            task2_in = [x for x in task2.children[4].children
+                        if isinstance(x, Reference)]
+            task2_out = [x for x in task2.children[5].children
+                         if isinstance(x, Reference)]
+
+            inout = itertools.product(task1_in, task2_out)
+            outin = itertools.product(task1_out, task2_in)
+            outout = itertools.product(task1_out, task2_out)
+            for mem in inout:
+                # Checking the symbol is the same works for non-structure
+                # References
+                if mem[0].symbol != mem[1].symbol:
+                    continue
+                # For structure reference we need to check they access
+                # the same member
+                if isinstance(mem[0], StructureReference):
+                    member = mem[0]
+                    member1 = mem[1]
+
+                    # We can't just do == on the Member child, as that
+                    # will recurse and check the array indices for any
+                    # ArrayMixin children
+
+                    # Get Members of each
+                    ref0_members = mem[0].walk(Member)
+                    ref1_members = mem[1].walk(Member)
+
+                    if len(ref0_members) != len(ref1_members):
+                        continue
+                    same_access = True
+                    for index, mem0 in enumerate(ref0_members):
+                        if mem0.name != ref1_members[index].name:
+                            same_access = False
+                    if not same_access:
+                        continue
+
+                # If we have (exactly) Reference objects we filter out
+                # non-matching ones with the symbol check, and matching ones
+                # are always valid since they are simple accesses.
+                if type(mem[0]) is Reference:
+                    continue
+
+                # All remaining objects are some sort of Array access
+                array1 = None
+                array2 = None
+                if isinstance(mem[0], ArrayReference):
+                    array1 = mem[0]
+                    array2 = mem[1]
+                else:
+                    array1 = mem[0].walk(ArrayMixin)[0]
+                    array2 = mem[1].walk(ArrayMixin)[0]
+                for i, index in enumerate(array1.indices):
+                    if isinstance(index, Literal):
+                        satisfiable = satisfiable and self._compare_literals(
```

Comment:
You don't, however those functions handle that automatically by checking both types are a Literal/Range as if either isn't the same type its bad.

I think the actual bad case is if `index` is not a Literal or Range and `array2.indices[i]` is one of those types. I think I assumed this would not fall over but I'm actually not sure.

Also I think a lot of this code is skippable whenever `satisfiable` is already false, so I think at the start of each of these loops I should have an 
`if not satisfiable: break` part as once we've identified that they're not satisfiable we should not bother to check any further for this pair because it can be quite expensive.

Reply:
Actually the middle paragraph here is fixed for the `outout` case, where it checks for if either is a `Literal` or Range`. I will copy this code to the other cases too.